# Homework 1 (Due Tuesday, March 22, 2022 at 6:29pm PST)

**Rubric**
* Identified 4 major themes from the reviews (2pts)
* Regex that groups / cleans the reviews is correctly implemented (4pts)
* Word count is correctly implemented (2pts)
* Analysis of recommendations and pitfalls/limitations are specific enough to be actionable (2pts)
* 

Not actionable recommendation:
* *The store managers should consider trying to improve the drive through experience to be more pleasant for customers*

More actionable recommendation:
* *Drive throughs are mentioned 23% of the time in reviews, and often focus on how slow the service is. We recommend adopting parallel drive through stations for Atlanta and Chicago*

You are a business analyst working for McDonalds. First, read through the reviews in `mcdonalds-yelp-negative-reviews.csv` (found in `datasets` folder). 

1. Identify 4 recurring themes/topics that reviewers are unhappy with. For example, one theme is that users are consistently unhappy with the drive-through experience.

2. Next, using regex, group together all occurrences of these phrases. For example, `drive-thru`, `drive through`, `drivethrough` can all be replaced as `_DRIVE_THROUGH_`.

3. Perform a word count, both overall, and broken out by city.

4. **Provide a few sentences with your findings and business recommendations.** Make any assumptions you'd like. I just want you to get into the habit of "finishing" your analysis: to avoid delivering technical numbers to a non-technical manager.

Some considerations in your analysis:

* Explain what some of **pitfalls/limitations** are of using only a word count analysis to make these inferences. What additional research/steps would you need to do to verify your conclusions?

**Submit everything as a new notebook and Slack direct message (group message) to me (Yu Chen) and the TAs (Mengqi Tan and Siyuan Ni) the HW as an attachment.**

**NOTE**: Name the notebook `lastname_firstname_HW1.ipynb`.

Every day late is -10%.

In [1]:
import pandas as pd
import re
from collections import Counter

In [2]:
text = open('../datasets/mcdonalds-yelp-negative-reviews.csv', 'r')
print(text)

<_io.TextIOWrapper name='../datasets/mcdonalds-yelp-negative-reviews.csv' mode='r' encoding='UTF-8'>


In [3]:
text.readline()

'_unit_id,city,review\n'

In [4]:
# text_df = pd.read_csv('../datasets/mcdonalds-yelp-negative-reviews.csv', encoding_errors='ignore')
text_df = pd.read_csv('../datasets/mcdonalds-yelp-negative-reviews.csv', encoding='latin1')
text_df

_unit_id         city  \
0     679455653      Atlanta   
1     679455654      Atlanta   
2     679455655      Atlanta   
3     679455656      Atlanta   
4     679455657      Atlanta   
...         ...          ...   
1520  679500008     Portland   
1521  679500224      Houston   
1522  679500608     New York   
1523  679501257      Chicago   
1524  679501402  Los Angeles   

                                                 review  
0     I'm not a huge mcds lover, but I've been to be...  
1     Terrible customer service. I came in at 9:30pm...  
2     First they "lost" my order, actually they gave...  
3     I see I'm not the only one giving 1 star. Only...  
4     Well, it's McDonald's, so you know what the fo...  
...                                                 ...  
1520  I enjoyed the part where I repeatedly asked if...  
1521  Worst McDonalds I've been in in a long time! D...  
1522  When I am really craving for McDonald's, this ...  
1523  Two points right out of the gate: 1. Thuggery ...  
1524  I wanted to grab breakfast one morning before ...  

[1525 rows x 3 columns]

<div class='alert-danger'>
    <p> Q:
    <p> how to tell when <i>text</i> and when <i>text_df</i>?
    <p> whether it's too much for memory?
    </div>

In [5]:
text_df.groupby('city')[['review']].count()

review
city               
Atlanta         130
Chicago         219
Cleveland        71
Dallas           75
Houston         105
Las Vegas       409
Los Angeles     167
New York        165
Portland         97

## task1

Identify 4 recurring themes/topics that reviewers are unhappy with.

In [6]:
def count_words(doc):
    counts = Counter()
    for r in doc:
        counts_tmp = Counter(re.findall(r'\w\w+', r, flags=re.IGNORECASE))
        counts += counts_tmp
    return counts

In [7]:
count_rough = count_words(text_df['review'])
# sorted(count_rough, key=lambda x: -count_rough[x])

In [8]:
count_rough.most_common(150)

[('the', 6237),
 ('and', 4137),
 ('to', 4030),
 ('of', 2005),
 ('is', 1918),
 ('was', 1793),
 ('in', 1788),
 ('it', 1756),
 ('for', 1651),
 ('this', 1427),
 ('my', 1421),
 ('that', 1327),
 ('you', 1252),
 ('they', 1232),
 ('at', 1019),
 ('have', 950),
 ('on', 908),
 ('not', 897),
 ('but', 858),
 ('me', 857),
 ('The', 845),
 ('order', 838),
 ('food', 832),
 ('McDonald', 822),
 ('with', 815),
 ('are', 707),
 ('one', 682),
 ('get', 665),
 ('there', 656),
 ('drive', 650),
 ('be', 648),
 ('so', 636),
 ('up', 586),
 ('here', 580),
 ('had', 566),
 ('just', 540),
 ('time', 525),
 ('or', 516),
 ('go', 511),
 ('out', 495),
 ('like', 483),
 ('service', 481),
 ('no', 468),
 ('as', 468),
 ('thru', 467),
 ('It', 464),
 ('place', 460),
 ('This', 455),
 ('when', 451),
 ('were', 445),
 ('McDonalds', 443),
 ('can', 423),
 ('your', 413),
 ('all', 405),
 ('only', 397),
 ('what', 396),
 ('if', 387),
 ('because', 382),
 ('location', 382),
 ('their', 373),
 ('we', 371),
 ('about', 366),
 ('been', 365),
 ('an

<div class="alert-info">
    Roughly speaking, there may be much complaint about <b>('food', 832), ('drive', 650), ('time', 525), ('service', 481), ('fries', 295), ('coffee', 254), ('breakfast', 174), ('wrong', 175), ('waiting', 153)</b> 
    </div>

<div class='alert-danger'>
    <p> TO-DO:
    <p> 1) how to tell <i>time</i> ~= <i>slow</i>? how to include <i>time</i> accurately in <i>slow</i> and exclude slow-irrelevant "time"s
    <p> 2) how to aggregate same topic? (e.g. slow, time, long, xx minutes)
    <p> 3) how to identify survivor bias? (e.g. people complaint fries more because they order fries more)
    </div>

<div class='alert-info'>
    <p> After scanning the reviews manually, 4 themes are identified below:
    <p> &emsp; - Drive Through
    <p> &emsp; - Customer Service (speed & attitude & correctness)
    <p> &emsp; - Food Quality (especially the fries, coffee and breakfast)
    <p> &emsp; - Dining Environment (dirty & noisy)
    </div>

## task2

Next, using regex, group together all occurrences of these phrases.

In [9]:
def standardize_word(doc, word_orig, word_std):
    for i in range(len(doc)):
#         rev_txt = rev.split(',"')[1] # it's for text not text_df
        rev = doc.iloc[i, -1]
        doc.iloc[i, -1] = re.sub(word_orig, word_std, rev, flags=re.IGNORECASE)

In [10]:
text_df['review_std'] = text_df['review']
text_df

_unit_id         city  \
0     679455653      Atlanta   
1     679455654      Atlanta   
2     679455655      Atlanta   
3     679455656      Atlanta   
4     679455657      Atlanta   
...         ...          ...   
1520  679500008     Portland   
1521  679500224      Houston   
1522  679500608     New York   
1523  679501257      Chicago   
1524  679501402  Los Angeles   

                                                 review  \
0     I'm not a huge mcds lover, but I've been to be...   
1     Terrible customer service. I came in at 9:30pm...   
2     First they "lost" my order, actually they gave...   
3     I see I'm not the only one giving 1 star. Only...   
4     Well, it's McDonald's, so you know what the fo...   
...                                                 ...   
1520  I enjoyed the part where I repeatedly asked if...   
1521  Worst McDonalds I've been in in a long time! D...   
1522  When I am really craving for McDonald's, this ...   
1523  Two points right out of the gate: 1. Thuggery ...   
1524  I wanted to grab breakfast one morning before ...   

                                             review_std  
0     I'm not a huge mcds lover, but I've been to be...  
1     Terrible customer service. I came in at 9:30pm...  
2     First they "lost" my order, actually they gave...  
3     I see I'm not the only one giving 1 star. Only...  
4     Well, it's McDonald's, so you know what the fo...  
...                                                 ...  
1520  I enjoyed the part where I repeatedly asked if...  
1521  Worst McDonalds I've been in in a long time! D...  
1522  When I am really craving for McDonald's, this ...  
1523  Two points right out of the gate: 1. Thuggery ...  
1524  I wanted to grab breakfast one morning before ...  

[1525 rows x 4 columns]

### Drive Through

<div class='alert-info'>
    drive-thru, drive through, drivethrough --> <b><i>_DRIVE_THROUGH_.
    </div>

In [11]:
word_orig, word_std = r'(drive-thru|drivethrough|drive through)', '_DRIVE_THROUGH_'
standardize_word(text_df, word_orig, word_std)

In [12]:
text_df

_unit_id         city  \
0     679455653      Atlanta   
1     679455654      Atlanta   
2     679455655      Atlanta   
3     679455656      Atlanta   
4     679455657      Atlanta   
...         ...          ...   
1520  679500008     Portland   
1521  679500224      Houston   
1522  679500608     New York   
1523  679501257      Chicago   
1524  679501402  Los Angeles   

                                                 review  \
0     I'm not a huge mcds lover, but I've been to be...   
1     Terrible customer service. I came in at 9:30pm...   
2     First they "lost" my order, actually they gave...   
3     I see I'm not the only one giving 1 star. Only...   
4     Well, it's McDonald's, so you know what the fo...   
...                                                 ...   
1520  I enjoyed the part where I repeatedly asked if...   
1521  Worst McDonalds I've been in in a long time! D...   
1522  When I am really craving for McDonald's, this ...   
1523  Two points right out of the gate: 1. Thuggery ...   
1524  I wanted to grab breakfast one morning before ...   

                                             review_std  
0     I'm not a huge mcds lover, but I've been to be...  
1     Terrible customer service. I came in at 9:30pm...  
2     First they "lost" my order, actually they gave...  
3     I see I'm not the only one giving 1 star. Only...  
4     Well, it's McDonald's, so you know what the fo...  
...                                                 ...  
1520  I enjoyed the part where I repeatedly asked if...  
1521  Worst McDonalds I've been in in a long time! D...  
1522  When I am really craving for McDonald's, this ...  
1523  Two points right out of the gate: 1. Thuggery ...  
1524  I wanted to grab breakfast one morning before ...  

[1525 rows x 4 columns]

### Customer Service
speed & attitude & correctness

#### speed

<div class='alert-info'>
    slow, slowly, sl-ow, SLOW, longer, ... --> <b><i>_CUSTOMER_SERVICE_SPEED_
    </div>

In [13]:
word_orig, word_std = r'(\bs(?:\-)?l(?:\-)?o(?:\-)?w(?:\-)?(?:ly)?\b|longer)', '_CUSTOMER_SERVICE_SPEED_'
standardize_word(text_df, word_orig, word_std)

<div class='alert-danger'>
    Q: any simplification?
    </div>

#### attitude

<div class='alert-info'>
    rude, unfriendly, unhappy, ... --> <b><i>_CUSTOMER_SERVICE_ATTITUDE_
    </div>

In [14]:
word_orig, word_std = r'(rude|unfriendly|unhappy)', '_CUSTOMER_SERVICE_ATTITUDE_'
standardize_word(text_df, word_orig, word_std)

#### correctness

<div class='alert-info'>
    wrong, wrongly ... --> <b><i>_CUSTOMER_SERVICE_CORRECTNESS_
    </div>

In [15]:
word_orig, word_std = r'(wrong(ly)?)', '_CUSTOMER_SERVICE_CORRECTNESS_'
standardize_word(text_df, word_orig, word_std)

### Food Quality 
especially the fries, coffee and breakfast

<div class='alert-info'>
    fries, coffee, breakfast, ... --> <b><i>_FOOD_QUALITY_
    </div>

In [16]:
word_orig, word_std = r'(fries|coffee|breakfast)', '_FOOD_QUALITY_'
standardize_word(text_df, word_orig, word_std)

### Dining Environment 
dirty & noisy

<div class='alert-info'>
    dirty, noisy, ... --> <b><i>_DINING_ENVIRONMENT_
    </div>

In [17]:
word_orig, word_std = r'(dirty|noisy)', '_DINING_ENVIRONMENT_'
standardize_word(text_df, word_orig, word_std)

## task3

Perform a word count, both overall, and broken out by city.

<div class='alert-info'>
    below func does absolute word count,
    </div>

In [18]:
# def count_certain_word_overall(doc, word):
#     counts = Counter()
#     for r in doc:
#         counts_tmp = Counter(re.findall(word, r))
#         counts += counts_tmp
#     return counts

<div class='alert-info'>
    word count on review level is preferred
    </div>

In [19]:
Counter(['burger', 'burger', 'burger'])

Counter({'burger': 3})

In [20]:
Counter(set(['burger', 'burger', 'burger']))

Counter({'burger': 1})

In [21]:
def count_certain_word_overall(doc, word):
    counts = Counter()
    for r in doc:
        counts_tmp = Counter(set(re.findall(word, r)))
        counts += counts_tmp
    return counts

In [82]:
def count_certain_word_bycity(df, word):
    result = pd.DataFrame(columns=['city', 'count'])
    for c in df['city'].unique():
        df_tmp = df.loc[df['city']==c]
        if df_tmp.empty: continue
#         counts = Counter()
#         for r in df_tmp['review_std']:
#             counts_tmp = Counter(re.findall(word, r))
#             counts += counts_tmp
        counts = count_certain_word_overall(df_tmp['review_std'], word)
        result = result.append({'city': c, 'count': list(counts.values())[0]}, ignore_index=True)
#         result = result.append({'city': c, 'count': counts.values()}, ignore_index=True)
    return result

### _DRIVE_THROUGH_

#### overall

In [84]:
w_std = '_DRIVE_THROUGH_'

In [85]:
count_drive = count_certain_word_overall(text_df['review_std'], w_std)
count_drive

Counter({'_DRIVE_THROUGH_': 222})

#### by city

In [86]:
count_drive_city = count_certain_word_bycity(text_df, w_std)
count_drive_city

city count
0      Atlanta    23
1    Las Vegas    65
2       Dallas    13
3     Portland    15
4      Chicago    22
5    Cleveland    12
6      Houston    15
7  Los Angeles    37
8     New York     1

### _CUSTOMER_SERVICE_

#### CUSTOMER_SERVICE

##### overall

In [87]:
w_std = '_CUSTOMER_SERVICE_'

In [88]:
count_serv = count_certain_word_overall(text_df['review_std'], w_std)
count_serv

Counter({'_CUSTOMER_SERVICE_': 387})

##### by city

In [89]:
count_serv_city = count_certain_word_bycity(text_df, w_std)
count_serv_city

city count
0      Atlanta    45
1    Las Vegas   109
2       Dallas    18
3     Portland    20
4      Chicago    55
5    Cleveland    27
6      Houston    25
7  Los Angeles    39
8     New York    29

#### '__SPEED_'

In [90]:
w_std = '_CUSTOMER_SERVICE_SPEED_'

In [91]:
count_serv_spd = count_certain_word_overall(text_df['review_std'], w_std)
count_serv_spd

Counter({'_CUSTOMER_SERVICE_SPEED_': 155})

In [92]:
count_serv_spd_city = count_certain_word_bycity(text_df, w_std)
count_serv_spd_city

city count
0      Atlanta    24
1    Las Vegas    42
2       Dallas     5
3     Portland     3
4      Chicago    24
5    Cleveland    10
6      Houston    14
7  Los Angeles    13
8     New York    13

#### '__ATTITUDE_'

In [93]:
w_std = '_CUSTOMER_SERVICE_ATTITUDE_'

In [94]:
count_serv_att = count_certain_word_overall(text_df['review_std'], w_std)
count_serv_att

Counter({'_CUSTOMER_SERVICE_ATTITUDE_': 128})

In [95]:
count_serv_att_city = count_certain_word_bycity(text_df, w_std)
count_serv_att_city

city count
0      Atlanta    15
1    Las Vegas    32
2       Dallas     6
3     Portland    11
4      Chicago    15
5    Cleveland    11
6      Houston     8
7  Los Angeles     9
8     New York    13

#### '__CORRECTNESS_'

In [96]:
w_std = '_CUSTOMER_SERVICE_CORRECTNESS_'

In [97]:
count_serv_cor = count_certain_word_overall(text_df['review_std'], w_std)
count_serv_cor

Counter({'_CUSTOMER_SERVICE_CORRECTNESS_': 155})

In [98]:
count_serv_cor_city = count_certain_word_bycity(text_df, w_std)
count_serv_cor_city

city count
0      Atlanta    14
1    Las Vegas    48
2       Dallas     9
3     Portland     8
4      Chicago    22
5    Cleveland     9
6      Houston    10
7  Los Angeles    22
8     New York     4

### _FOOD_QUALITY_

#### overall

In [99]:
w_std = '_FOOD_QUALITY_'

In [100]:
count_food = count_certain_word_overall(text_df['review_std'], w_std)
count_food

Counter({'_FOOD_QUALITY_': 455})

#### by city

In [101]:
count_food_city = count_certain_word_bycity(text_df, w_std)
count_food_city

city count
0      Atlanta    41
1    Las Vegas   133
2       Dallas    15
3     Portland    27
4      Chicago    64
5    Cleveland    13
6      Houston    33
7  Los Angeles    52
8     New York    52

### _DINING_ENVIRONMENT_

#### overall

In [102]:
w_std = '_DINING_ENVIRONMENT_'

In [103]:
count_env = count_certain_word_overall(text_df['review_std'], w_std)
count_env

Counter({'_DINING_ENVIRONMENT_': 64})

#### by city

In [104]:
count_env_city = count_certain_word_bycity(text_df, w_std)
count_env_city

city count
0      Atlanta     8
1    Las Vegas    17
2       Dallas     2
3     Portland     3
4      Chicago     5
5    Cleveland     6
6      Houston     5
7  Los Angeles     7
8     New York     6

### AGGREGATION

In [105]:
count_drive_city.set_index('city', inplace=True)
count_drive_city.columns = ['drive']
count_env_city.set_index('city', inplace=True)
count_env_city.columns = ['env']
count_food_city.set_index('city', inplace=True)
count_food_city.columns = ['food']
count_serv_city.set_index('city', inplace=True)
count_serv_city.columns = ['serve']
count_serv_att_city.set_index('city', inplace=True)
count_serv_att_city.columns = ['attitude']
count_serv_spd_city.set_index('city', inplace=True)
count_serv_spd_city.columns = ['speed']
count_serv_cor_city.set_index('city', inplace=True)
count_serv_cor_city.columns = ['correctness']

In [126]:
result = pd.concat([count_drive_city, count_serv_city, count_serv_att_city, count_serv_spd_city,
                    count_serv_cor_city, count_food_city, count_env_city], axis=1)
result

drive serve attitude speed correctness food env
city                                                       
Atlanta        23    45       15    24          14   41   8
Las Vegas      65   109       32    42          48  133  17
Dallas         13    18        6     5           9   15   2
Portland       15    20       11     3           8   27   3
Chicago        22    55       15    24          22   64   5
Cleveland      12    27       11    10           9   13   6
Houston        15    25        8    14          10   33   5
Los Angeles    37    39        9    13          22   52   7
New York        1    29       13    13           4   52   6

In [127]:
result['sum_city'] = result.sum(axis=1)
result.sort_values('sum_city', ascending=False)

drive serve attitude speed correctness food env  sum_city
city                                                                 
Las Vegas      65   109       32    42          48  133  17     446.0
Chicago        22    55       15    24          22   64   5     207.0
Los Angeles    37    39        9    13          22   52   7     179.0
Atlanta        23    45       15    24          14   41   8     170.0
New York        1    29       13    13           4   52   6     118.0
Houston        15    25        8    14          10   33   5     110.0
Cleveland      12    27       11    10           9   13   6      88.0
Portland       15    20       11     3           8   27   3      87.0
Dallas         13    18        6     5           9   15   2      68.0

In [132]:
sum_comp = result.sum(axis=0)
sum_comp.sort_values(ascending=False)

sum_city       1473.0
food              430
serve             367
drive             203
speed             148
correctness       146
attitude          120
env                59
dtype: object

## task4

business recommendations & pitfalls/limitations

### business recommendations
Provide a few sentences with your findings and business recommendations.

In [120]:
result.loc['Las Vegas', 'sum_city'] / sum(result['sum_city'])

0.3027834351663272

In [122]:
result.loc['Las Vegas', 'food'] / result.loc['Las Vegas', 'sum_city']

0.2982062780269058

In [121]:
result.loc['Las Vegas', 'serve'] / result.loc['Las Vegas', 'sum_city']

0.24439461883408073

<div class='alert-info'>
    <b>Recommendation #1:</b>
Overall, Las Vegas has most complaints, accounting for 30% of the total. Among the complaints, 30% are about the food quality and 24% are about customer service, both of which are employee-related. Therefore, McDonalds should provide extra training for employees in Las Vegas.
</div>

In [133]:
sum_comp['food'] / sum_comp['sum_city']

0.2919212491513917

<div class='alert-info'>
    <b>Recommendation #2:</b>
Among the four topics, food quality complaints rank the 1st and account for 30% of the total. In this analysis, we only detect complaints about fries, coffee and breakfast, yet there're considerable complaints already. McDonalds should increase their standard for the food quality and require an extra examination before handing over the food.
</div>

In [135]:
sum_comp['speed'] /sum_comp['serve'], \
sum_comp['correctness'] /sum_comp['serve'], \
sum_comp['attitude'] /sum_comp['serve'] 

(0.4032697547683924, 0.3978201634877384, 0.32697547683923706)

<div class='alert-info'>
    <b>Recommendation #3:</b>
Customer service is an important element of McDonalds' value chain. In this analysis, 40% complaints are about service speed, 40% are about service correctness and 33% are about service attitude. McDonalds should consider a more strict employee training program and/or increase the number of employees per location, especially for busy cities like Las Vegas, Chicago and Atlanta.
</div>

### pitfalls/limitations
Explain what some of pitfalls/limitations are of using only a word count analysis to make these inferences. What additional research/steps would you need to do to verify your conclusions?

<div class='alert-info'>
    <p><b>Pitfalls/Limitations:</b>
In this case, word count analysis are not confirmable since we ignore the context of the words. For example, count of <i>slow</i> may take <i>not slow</i> into account and thus lead to false positives. Also, it is hard to consider all synonyms of certain topic given the variety of expressions, which would lead to false negatives.
    <p><b>Additional research/steps:</b>
Analysis that take context into consideration would be necessary for this NLP case. My analysis here only offers rough conclusions and without advanced NLP techniques, manual analysis is inevitable.
</div>